internal/authorization/authorization.go

```diff
@@ -0,0 +1,85 @@
+package authorization
+
+import (
+	"context"
+	"database/sql"
+	"net/http"
+
+	"Twitter_like_application/internal/services"
+)
+
+const (
+	ctxKeyUserID  = "userID"
+	ctxKeyIsAdmin = "isAdmin"
+)
+
+type Service struct {
+	DB *sql.DB
+}
+
+func New(db *sql.DB) *Service {
+	return &Service{DB: db}
+}
+
+func (s *Service) CheckAuth(w http.ResponseWriter, r *http.Request) *http.Request {
+	apikey := r.Header.Get("X-API-KEY")
+	cookie, err := r.Cookie("session")
+	if apikey == "" && (err != nil || cookie == nil) {
+		services.ReturnErr(w, "Unauthorized", http.StatusUnauthorized)
+		return nil
+	}
+	var sessionID string
+	if apikey != "" {
+		sessionID = apikey
+	} else if cookie != nil {
+		sessionID = cookie.Value
+	}
+	if sessionID == "" {
+		services.ReturnErr(w, "Unauthorized", http.StatusUnauthorized)
+		return nil
+	}
+	query := `SELECT us.user_id, ut.admin FROM user_session us JOIN users_tweeter ut ON us.user_id = ut.id WHERE us.session_id = $1 LIMIT 1`
+
+	var userID int
+	var isAdmin bool
+	err = s.DB.QueryRow(query, sessionID).Scan(&userID, &isAdmin)
+	if err != nil {
+		services.ReturnErr(w, err.Error(), http.StatusInternalServerError)
+		return nil
+	}
+	ctx := context.WithValue(r.Context(), ctxKeyUserID, userID)
+	ctx = context.WithValue(ctx, ctxKeyIsAdmin, isAdmin)
+	return r.WithContext(ctx)
+}
+
+func (s *Service) AuthHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		r = s.CheckAuth(w, r)
+		if v, ok := r.Context().Value(ctxKeyUserID).(string); ok && v != "" {
+			next.ServeHTTP(w, r)
+		} else {
+			services.ReturnErr(w, "Unauthorized", http.StatusUnauthorized)
+		}
+
+	})
+}
+func (s *Service) UnauthenticatedHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		newR := s.CheckAuth(w, r)
+		if newR == nil {
+			newR = r
+		}
```

Comment:
this looks backwards. It would be better to create a method that would just fill-in the context and always return a request, and use it everywhere